Layout basic quiz React components on /app/#/quiz

phorust opened the pull request:
Placed the quiz components in app/components/quiz. It renders some mock data found in a/c/q/Quiz.js and has a SASS file associated with it.

I also added the jquery functionality back into the topic picker that I made a while back, using the proper webpack jquery. You can once again change topics and pin the topic picker so it doesn't disappear.

btburton42 commented on app/components/quiz/Quiz.js:
```diff
@@ -0,0 +1,63 @@
+  },
+
+  getDefaultProps: function() {
+    return {
```
ideally i'd like to see this worked into an action so that we can be prepared to consume actual requests when the time comes. or at least a #TODO comment for the time being.

mngyuan replied:
Agreed - I need to get familiar with our actions and store setup first

btburton42 commented on app/components/quiz/Quiz.js:
```diff
@@ -0,0 +1,63 @@
+  },
+
+  onUpdate: function(value) {
+    this.setState({value});
```
you should use the actions --> store methods to push updates to state. more on that api here http://alt.js.org/docs/actions/ but i can pseudo-code this in if the documentation isn't clear.

btburton42 replied:
also i'm wondering if i shouldn't move this to redux.js as i've heard good things.

mngyuan replied:
i've heard amazing things about redux - wouldn't be opposed to moving to that. i'll change this onUpdate handler to use actions / the store for now

btburton42 commented on app/components/quiz/Quiz.js:
```diff
@@ -0,0 +1,63 @@
+                          onUpdate={this.onUpdate} />
+          );
+        })}
+        <button className="quiz__next" {...opts}>Next</button>
```
same thing here. we should refactor in with alt.

btburton42 commented on app/components/quiz/QuizContext.js:
```diff
@@ -0,0 +1,43 @@
+    return (
+      <p className="quiz__context">
+        {highlights.map((n, i) => {
+          // I'd rather not have to wrap these like this... is there a better way?
```
this is a clue that it should be abstracted into another component.

btburton42 replied:
oh and if you mean wrapping the span i think that's because react expects a div or a component so wrapping it in parenthesis allows it to wrap this in a div. do you see that in the markup?

mngyuan replied:
oh, I'm referring to having

```
<span>
  <span />
  <span />
</span>
```

as opposed to being able to return two `<span>` as peers. it's a minor thing but the resulting markup on the page would be cleaner - I think there's a way to iterate so I can achieve this, but I couldn't figure out how without making my iteration significantly uglier.

anywho won't worry about it for now

btburton42 commented on app/components/quiz/QuizAnswer.js:
```diff
@@ -0,0 +1,35 @@
+import React from 'react';
+import AppStore from 'store/appStore';
```
this isn't doing anything i can see. but we should be pushing value changes to the answer input into state via the actions --> store.

btburton42 commented on app/components/quiz/QuizContext.js:
```diff
@@ -0,0 +1,43 @@
+    var text = this.props.context.text;
+    var highlights = this.props.context.highlights;
+
+    var start = 0;
```
what is this block through L25 doing?

mngyuan replied:
it takes care of if there's a tail after the highlighted portions which needs to be displayed i.e.

```
highlights = [ [0,5] ]
text = "01234this is a tail which is not highlighted so it will be shown in a different color"
```

actually thanks for asking I just realized there's a bug with the way I'm iterating

btburton42 commented on app/components/topicPicker.js:
```diff
@@ -2,6 +2,7 @@ import React from 'react';
 import {RouteHandler} from 'react-router';
 import objectAssign from 'object-assign';
 import ListenerMixin from 'alt/mixins/ListenerMixin';
+import jquery from 'jquery';
```
i think you can do `import $ from 'jquery'` if i'm not mistaken.

mngyuan replied:
:+1:

mngyuan replied:
actually, eslint complains since `$ is already declared in the upper scope`.

someone else is declaring `$`. I'll just stick to `var $ = jquery` later in the code.

btburton42 commented on app/styles/_common.scss:
```diff
@@ -0,0 +1,10 @@
+}
+
+#react-root, .app, body, html {
+  width: 100%;
```
do we need this?

mngyuan replied:
well yes the way I've been positioning `<Quiz>` in the page is vertical-aligning it center, which requires width and height 100%... but that layout seems a bit overkill or it'll be annoying when the questions and their answers become longer than the viewport.

i'll swap this out for something simpler.

btburton42 commented on app/styles/quiz.scss:
```diff
@@ -0,0 +1,34 @@
+    margin-top: 20px;
+    float: right;
+    font-family: $font-ui;
+    font-weight: 600;
```
i just use bold but that's me. :)

mngyuan replied:
oh, that's be way more readable. changing

btburton42 commented on app/components/quiz/QuizQuestion.js:
```diff
@@ -0,0 +1,38 @@
+  },
+
+  onChange: function(e) {
+    // doesn't handle checkboxes
```
do you meant this only handles text inputs?

mngyuan replied:
oh - it only handles how to update the `<Quiz>`'s answer if you're using radio buttons. commit coming with checkbox support

phorust commented:
hey Tyler, so haven't put in any of the store or action stuff - can we split that into a separate task, and merge this? that way @arjunmehta94 can work off master instead of a branch based on this PR. i cleaned up some of the CSS and fixed a silly iteration bug, and made it so the `<Quiz>` can correctly store the answer state (as an array) when checkboxes are used instead of radio buttons.

btburton42 commented:
Let me take a quick look.

On Wed, Feb 3, 2016 at 4:23 PM Kevin Lee <notifications@github.com> wrote:

> hey Tyler, so haven't put in any of the store or action stuff - can we
> split that into a separate task, and merge this? that way @arjunmehta94
> <https://github.com/arjunmehta94> can work off master instead of a branch
> based on this PR. i cleaned up some of the CSS and fixed a silly iteration
> bug, and made it so the <Quiz> can correctly store the answer state (as
> an array) when checkboxes are used instead of radio buttons.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Goodly/annotator-content-analysis/pull/16#issuecomment-179549046>
> .
>
-- 
--
Tyler Burton
btburton42@gmail.com

btburton42 commented:
there's a few nitpicky bugs in here and some stuff i know you're working on, but since you guys are in quick dev cycles let's give this a go and see what comes out of the wash in the next PRs... 

:+1:

phorust commented:
alright, let's ship it
![](http://i.giphy.com/wpoLqr5FT1sY0.gif)

nickbadams commented:
👍

On Thu, Feb 4, 2016 at 2:47 PM, Kevin Lee <notifications@github.com> wrote:

> alright, let's ship it
>
> <https://camo.githubusercontent.com/b65c3052e7bd565e14e143e601a9dabc713d0e18/687474703a2f2f692e67697068792e636f6d2f77706f4c7172354654317359302e676966>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Goodly/annotator-content-analysis/pull/16#issuecomment-180087823>
> .
>



-- 

*Nick Adams, Ph.D.Data Science Fellow, *

*Berkeley Institute for Data Science*
*University of California, Berkeley*
